### PR TITLE
operator: multiarch adaptor image

### PR DESCRIPTION
**Pull request by jtumber-ibm**
- Build cross-compiled cloud-api-adaptor in builder step of Dockerfile.
- Update build script to fully utilise buildx

Test run: https://hub.docker.com/r/jamesnelson/cloud-api-adaptor-ibmcloud/tags

Fixes: #533

Signed-off-by: James Tumber <james.tumber@ibm.com>

### CODE REVIEW

**Comment by liudalibj** [comment]
I built out the multi-arch image, but it seems that the `cloud-api-adaptor-ibmcloud` binary in s390x docker images is bad.
```
root@csi-x86-worker:~/cloud-api-adaptor# export registry=liudali
root@csi-x86-worker:~/cloud-api-adaptor# CLOUD_PROVIDER=ibmcloud make image
hack/build.sh
~/cloud-api-adaptor ~/cloud-api-adaptor
[+] Building 9.5s (22/22) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                                                            0.0s
 => => transferring dockerfile: 846B                                                                                                                                                                                                            0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                               0.0s
 => => transferring context: 2B                                                                                                                                                                                                                 0.0s
 => [linux/s390x internal] load metadata for docker.io/library/fedora:36                                                                                                                                                                        0.6s
 => [linux/amd64 internal] load metadata for docker.io/library/golang:1.18                                                                                                                                                                      0.5s
 => [linux/amd64 internal] load metadata for docker.io/library/fedora:36                                                                                                                                                                        0.5s
 => [linux/amd64 stage-1 1/4] FROM docker.io/library/fedora:36@sha256:17ab257e353f7e0247ce4b1a6ac7a4dd372cf217e532f27d67feb876a53af471                                                                                                          0.0s
 => => resolve docker.io/library/fedora:36@sha256:17ab257e353f7e0247ce4b1a6ac7a4dd372cf217e532f27d67feb876a53af471                                                                                                                              0.0s
 => [internal] load build context                                                                                                                                                                                                               0.4s
 => => transferring context: 696.67kB                                                                                                                                                                                                           0.4s
 => [linux/s390x stage-1 1/4] FROM docker.io/library/fedora:36@sha256:17ab257e353f7e0247ce4b1a6ac7a4dd372cf217e532f27d67feb876a53af471                                                                                                          0.0s
 => => resolve docker.io/library/fedora:36@sha256:17ab257e353f7e0247ce4b1a6ac7a4dd372cf217e532f27d67feb876a53af471                                                                                                                              0.0s
 => [linux/amd64 builder 1/6] FROM docker.io/library/golang:1.18@sha256:50c889275d26f816b5314fc99f55425fa76b18fcaf16af255f5d57f09e1f48da                                                                                                        0.0s
 => => resolve docker.io/library/golang:1.18@sha256:50c889275d26f816b5314fc99f55425fa76b18fcaf16af255f5d57f09e1f48da                                                                                                                            0.0s
 => CACHED [linux/s390x stage-1 2/4] RUN if [ "ibmcloud" = "libvirt" ] ; then dnf install -y libvirt-libs genisoimage /usr/bin/ssh && dnf clean all; fi                                                                                         0.0s
 => CACHED [linux/amd64 builder 2/6] COPY . cloud-api-adaptor                                                                                                                                                                                   0.0s
 => CACHED [linux/amd64 builder 3/6] RUN git clone -b CCv0 https://github.com/kata-containers/kata-containers                                                                                                                                   0.0s
 => CACHED [linux/amd64 builder 4/6] WORKDIR cloud-api-adaptor                                                                                                                                                                                  0.0s
 => CACHED [linux/amd64 builder 5/6] RUN if [ "ibmcloud" = "libvirt" ] ; then apt-get update -y && apt-get install -y libvirt-dev; fi                                                                                                           0.0s
 => CACHED [linux/amd64 builder 6/6] RUN ARCH=${TARGETARCH} make                                                                                                                                                                                0.0s
 => CACHED [linux/s390x stage-1 3/4] COPY --from=builder /go/cloud-api-adaptor/cloud-api-adaptor /usr/local/bin/cloud-api-adaptor-ibmcloud                                                                                                      0.0s
 => CACHED [linux/s390x stage-1 4/4] COPY --from=builder /go/cloud-api-adaptor/entrypoint.sh /usr/local/bin/entrypoint.sh                                                                                                                       0.0s
 => CACHED [linux/amd64 stage-1 2/4] RUN if [ "ibmcloud" = "libvirt" ] ; then dnf install -y libvirt-libs genisoimage /usr/bin/ssh && dnf clean all; fi                                                                                         0.0s
 => CACHED [linux/amd64 stage-1 3/4] COPY --from=builder /go/cloud-api-adaptor/cloud-api-adaptor /usr/local/bin/cloud-api-adaptor-ibmcloud                                                                                                      0.0s
 => CACHED [linux/amd64 stage-1 4/4] COPY --from=builder /go/cloud-api-adaptor/entrypoint.sh /usr/local/bin/entrypoint.sh                                                                                                                       0.0s
 => exporting to image                                                                                                                                                                                                                          8.3s
 => => exporting layers                                                                                                                                                                                                                         0.0s
 => => exporting manifest sha256:06f0173a9f7eeae17831d36df5c3f93f69fd3e0545cdd3e06b2f121b60eeb6e2                                                                                                                                               0.0s
 => => exporting config sha256:64a6a83b0756f64ad3b8919379d12c2beaf5d1309837ab741ef21b17891be12f                                                                                                                                                 0.0s
 => => exporting manifest sha256:4c883161ae6525d2b0ba254c88a676dc9b584f19b1369fbcd5167ff3049bdf1f                                                                                                                                               0.0s
 => => exporting config sha256:aa532f94a09241b8576c4c4994ca73e7a9fbae00f12226a2db2c288cec57c771                                                                                                                                                 0.0s
 => => exporting manifest list sha256:97f23dbc7affe749a4adf676a21fdfac5b1e41ec3917f44497420e239cb93fc1                                                                                                                                          0.0s
 => => pushing layers                                                                                                                                                                                                                           6.8s
 => => pushing manifest for docker.io/liudali/cloud-api-adaptor-ibmcloud:2023011814421674052949@sha256:97f23dbc7affe749a4adf676a21fdfac5b1e41ec3917f44497420e239cb93fc1                                                                         1.5s
 => [auth] liudali/cloud-api-adaptor-ibmcloud:pull,push token for registry-1.docker.io                                                                                                                                                          0.0s
~/cloud-api-adaptor
```
- Check the `cloud-api-adaptor` binary from s390x docker image
```
root@peer-pod-z-csi-worker:~# /root/cloud-api-adaptor/ibmcloud/image/skopeo/bin/skopeo copy --insecure-policy docker://docker.io/liudali/cloud-api-adaptor-ibmcloud:2023011814421674052949 oci:/tmp/test_s390x:latest
Getting image source signatures
Copying blob ae0977f45148 done
Copying blob 4f4fb700ef54 done
Copying blob 6c103a2964bd done
Copying blob f3a7ed020550 done
Copying config aa4dccc807 done
Writing manifest to image destination
Storing signatures

root@peer-pod-z-csi-worker:~# /root/cloud-api-adaptor/ibmcloud/image/umoci/umoci unpack --image /tmp/test_s390x:latest /tmp/s390x_bundle
root@peer-pod-z-csi-worker:~# /tmp/s390x_bundle/rootfs/usr/local/bin/cloud-api-adaptor-ibmcloud
-bash: /tmp/s390x_bundle/rootfs/usr/local/bin/cloud-api-adaptor-ibmcloud: cannot execute binary file: Exec format error
root@peer-pod-z-csi-worker:~# file /tmp/s390x_bundle/rootfs/usr/local/bin/cloud-api-adaptor-ibmcloud
/tmp/s390x_bundle/rootfs/usr/local/bin/cloud-api-adaptor-ibmcloud: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, not stripped
root@peer-pod-z-csi-worker:~#
```

Caused by docker cache?

**Review by liudalibj — CHANGES_REQUESTED** [review]
Need fix the cached issue.

**Comment by jtumber-ibm** [comment]
> Need fix the cached issue.

Thanks for taking a look, this is very strange for me it works fine even with caching enabled. I can add `--no-cache` flag to prevent it happening.

**Comment by liudalibj** [comment]
I tried add the `--no-cache` option to the docker buildx command the issue not be fixed.
```
hack/build.sh
~/cloud-api-adaptor ~/cloud-api-adaptor
[+] Building 116.4s (22/22) FINISHED
 => [internal] load .dockerignore                                                                                                                                                                                                               0.0s
 => => transferring context: 2B                                                                                                                                                                                                                 0.0s
 => [internal] load build definition from Dockerfile                                                                                                                                                                                            0.0s
 => => transferring dockerfile: 846B                                                                                                                                                                                                            0.0s
 => [linux/s390x internal] load metadata for docker.io/library/fedora:36                                                                                                                                                                        0.6s
 => [linux/amd64 internal] load metadata for docker.io/library/fedora:36                                                                                                                                                                        0.5s
 => [linux/amd64 internal] load metadata for docker.io/library/golang:1.18                                                                                                                                                                      0.6s
 => [internal] load build context                                                                                                                                                                                                               0.6s
 => => transferring context: 696.67kB                                                                                                                                                                                                           0.5s
 => CACHED [linux/amd64 builder 1/6] FROM docker.io/library/golang:1.18@sha256:50c889275d26f816b5314fc99f55425fa76b18fcaf16af255f5d57f09e1f48da                                                                                                 0.0s
 => => resolve docker.io/library/golang:1.18@sha256:50c889275d26f816b5314fc99f55425fa76b18fcaf16af255f5d57f09e1f48da                                                                                                                            0.0s
 => CACHED [linux/s390x stage-1 1/4] FROM docker.io/library/fedora:36@sha256:17ab257e353f7e0247ce4b1a6ac7a4dd372cf217e532f27d67feb876a53af471                                                                                                   0.1s
 => => resolve docker.io/library/fedora:36@sha256:17ab257e353f7e0247ce4b1a6ac7a4dd372cf217e532f27d67feb876a53af471                                                                                                                              0.1s
 => CACHED [linux/amd64 stage-1 1/4] FROM docker.io/library/fedora:36@sha256:17ab257e353f7e0247ce4b1a6ac7a4dd372cf217e532f27d67feb876a53af471                                                                                                   0.1s
 => => resolve docker.io/library/fedora:36@sha256:17ab257e353f7e0247ce4b1a6ac7a4dd372cf217e532f27d67feb876a53af471                                                                                                                              0.1s
 => [linux/amd64 stage-1 2/4] RUN if [ "ibmcloud" = "libvirt" ] ; then dnf install -y libvirt-libs genisoimage /usr/bin/ssh && dnf clean all; fi                                                                                                0.4s
 => [linux/s390x stage-1 2/4] RUN if [ "ibmcloud" = "libvirt" ] ; then dnf install -y libvirt-libs genisoimage /usr/bin/ssh && dnf clean all; fi                                                                                                0.5s
 => [linux/amd64 builder 2/6] COPY . cloud-api-adaptor                                                                                                                                                                                         34.1s
 => [linux/amd64 builder 3/6] RUN git clone -b CCv0 https://github.com/kata-containers/kata-containers                                                                                                                                          8.7s
 => [linux/amd64 builder 4/6] WORKDIR cloud-api-adaptor                                                                                                                                                                                         0.1s
 => [linux/amd64 builder 5/6] RUN if [ "ibmcloud" = "libvirt" ] ; then apt-get update -y && apt-get install -y libvirt-dev; fi                                                                                                                  0.1s
 => [linux/amd64 builder 6/6] RUN ARCH=${TARGETARCH} make                                                                                                                                                                                      61.1s
 => [linux/s390x stage-1 3/4] COPY --from=builder /go/cloud-api-adaptor/cloud-api-adaptor /usr/local/bin/cloud-api-adaptor-ibmcloud                                                                                                             1.5s
 => [linux/amd64 stage-1 3/4] COPY --from=builder /go/cloud-api-adaptor/cloud-api-adaptor /usr/local/bin/cloud-api-adaptor-ibmcloud                                                                                                             1.4s
 => [linux/amd64 stage-1 4/4] COPY --from=builder /go/cloud-api-adaptor/entrypoint.sh /usr/local/bin/entrypoint.sh                                                                                                                              0.1s
 => [linux/s390x stage-1 4/4] COPY --from=builder /go/cloud-api-adaptor/entrypoint.sh /usr/local/bin/entrypoint.sh                                                                                                                              0.1s
 => exporting to image                                                                                                                                                                                                                          8.9s
 => => exporting layers                                                                                                                                                                                                                         2.2s
 => => exporting manifest sha256:5f09eab6330187c4f6a21ae8c709e385f5214dd468d6195006897b673b330361                                                                                                                                               0.0s
 => => exporting config sha256:27bad734ea369242096e6edf50a7fb792c0f5dd8d0175e452535f5ddb70e4f8a                                                                                                                                                 0.0s
 => => exporting manifest sha256:879293f46fd574bfc377131a6d0de0e7263c1e11e67c3f8c5a34de7ba8e08542                                                                                                                                               0.0s
 => => exporting config sha256:afef96dbeccbb41dd877dba1d0bf9655c9eca5c0d1cf9212f57f24debeefcd30                                                                                                                                                 0.0s
 => => exporting manifest list sha256:36f907b292281569cd9ab009f148bb17e584b289bcf88143a2f8b2c0a616def3                                                                                                                                          0.0s
 => => pushing layers                                                                                                                                                                                                                           5.2s
 => => pushing manifest for docker.io/liudali/cloud-api-adaptor-ibmcloud:2023011815001674054043@sha256:36f907b292281569cd9ab009f148bb17e584b289bcf88143a2f8b2c0a616def3                                                                         1.5s
 => [auth] liudali/cloud-api-adaptor-ibmcloud:pull,push token for registry-1.docker.io                                                                                                                                                          0.0s
~/cloud-api-adaptor
```

From the logs
```
 => [linux/s390x stage-1 3/4] COPY --from=builder /go/cloud-api-adaptor/cloud-api-adaptor /usr/local/bin/cloud-api-adaptor-ibmcloud                                                                                                             1.5s
 => [linux/amd64 stage-1 3/4] COPY --from=builder /go/cloud-api-adaptor/cloud-api-adaptor /usr/local/bin/cloud-api-adaptor-ibmcloud  
```
is`-from=builder /go/cloud-api-adaptor/cloud-api-adaptor` using the same file?

**Comment by jtumber-ibm** [comment]
@liudalibj issue is resolved I think, was missing the `ARG TARGETARCH` line to expose the variable to the `RUN` command. `RUN ARCH=$TARGETARCH make` was being always run as `RUN ARCH= make`. 

Checking my image `docker pull jamesnelson/cloud-api-adaptor-ibmcloud:2023011816141674058450` using your instructions

```
skopeo --override-arch=s390x copy docker://docker.io/jamesnelson/cloud-api-adaptor-ibmcloud:2023011816141674058450 oci:/tmp/test_s390x:latest
./podvm/umoci/umoci unpack --rootless --image /tmp/test_s390x:latest /tmp/s390x_bundle
file /tmp/s390x_bundle/rootfs/usr/local/bin/cloud-api-adaptor-ibmcloud 
/tmp/s390x_bundle/rootfs/usr/local/bin/cloud-api-adaptor-ibmcloud: ELF 64-bit MSB executable, IBM S/390, version 1 (SYSV), statically linked, Go BuildID=558de2NadQP2M9jOqyhp/SNHkjM7IKRaESpByfrqL/3_RNXnFQdh85Sfe0Kuvg/dNkYhobV2E6e4yRr4EeO, with debug_info, not stripped
```

**Review by liudalibj — APPROVED** [review]
Rebuild the image with the update PR, the images work as expected.

The PR is LGTM

**Review by stevenhorsman — APPROVED** [review]
LGTM
I ran:
```
# export registry=stevenhorsman
# CLOUD_PROVIDER=ibmcloud make image
```
on an x86 machine, which produced the image https://hub.docker.com/layers/stevenhorsman/cloud-api-adaptor-ibmcloud/2023011909301674120612/images/sha256-bce0b57e3adeb81339b012e5cf0bbd9ab24d09bcd120ba3ead073727808df48a?context=repo.
I then tried pulling it down and unpacking on a s390x machine and checking the file type which worked:
```
# /root/cloud-api-adaptor/ibmcloud/image/skopeo/bin/skopeo copy --insecure-policy --src-creds ${SOURCE_CREDS} docker://docker.io/stevenhorsman/cloud-api-adaptor-ibmcloud:2023011909301674120612 oci:/tmp/test_s390x:latest
Getting image source signatures
Copying blob ae0977f45148 skipped: already exists
Copying blob 4f4fb700ef54 skipped: already exists
Copying blob 6756917e4e79 done
Copying blob a310eff2a09e done
Copying config a69b09f992 done
Writing manifest to image destination
Storing signatures
root@sh-pp-z-cluster-2023-01-18-worker:~# rm -rf /tmp/s390x_bundle/ && /root/cloud-api-adaptor/ibmcloud/image/umoci/umoci unpack --image /tmp/test_s390x:latest /tmp/s390x_bundle
root@sh-pp-z-cluster-2023-01-18-worker:~# file /tmp/s390x_bundle/rootfs/usr/local/bin/cloud-api-adaptor-ibmcloud
/tmp/s390x_bundle/rootfs/usr/local/bin/cloud-api-adaptor-ibmcloud: ELF 64-bit MSB executable, IBM S/390, version 1 (SYSV), statically linked, not stripped
root@sh-pp-z-cluster-2023-01-18-worker:~#
```
and did the same on an x86 machine, which also worked:
```
# /root/cloud-api-adaptor/ibmcloud/image/skopeo/bin/skopeo copy --insecure-policy --src-creds ${SOURCE_CREDS} docker://stevenhorsman/cloud-api-adaptor-ibmcloud:2023011909301674120612 oci:/tmp/test_amd64:latest
Getting image source signatures
Copying blob c13ad3937d91 skipped: already exists
Copying blob 4f4fb700ef54 skipped: already exists
Copying blob 7af247f5bef3 skipped: already exists
Copying blob f6a3ce658288 skipped: already exists
Copying config 8b49cd62c2 done
Writing manifest to image destination
Storing signatures
root@sh-pp-cluster-2023-01-17-worker:~/jt-caa# /root/cloud-api-adaptor/ibmcloud/image/umoci/umoci unpack --image /tmp/test_amd64:latest /tmp/amd64_bundle
   ⨯ create runtime bundle: config.json already exists in /tmp/amd64_bundle
root@sh-pp-cluster-2023-01-17-worker:~/jt-caa# rm -rf /tmp/amd64_bundle/ && !!
rm -rf /tmp/amd64_bundle/ && /root/cloud-api-adaptor/ibmcloud/image/umoci/umoci unpack --image /tmp/test_amd64:latest /tmp/amd64_bundle
root@sh-pp-cluster-2023-01-17-worker:~/jt-caa# file /tmp/amd64_bundle/rootfs/usr/local/bin/cloud-api-adaptor-ibmcloud
/tmp/amd64_bundle/rootfs/usr/local/bin/cloud-api-adaptor-ibmcloud: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, Go BuildID=og3AhdinPWhzRJ4PKGOq/aFZIW1bMd7q6KhHw5Q9f/xFjKqkbHkMUazCRYdQPE/NC9azHX4BGAlTZE6j8p7, not stripped
```

Thanks @jtumber-ibm !

File: hack/build.sh [hunk 1]
```diff
@@ -6,58 +6,27 @@ set -o nounset
...
-	docker manifest create \
-		${registry}:latest \
-		--amend ${registry}:x86_64-${tag}
-
-	docker manifest push ${registry}:${tag}
-	docker manifest push --purge ${registry}:latest
+	docker buildx build --platform "${supported_arches}" \
+		--build-arg CLOUD_PROVIDER="${CLOUD_PROVIDER}" \
+		-f Dockerfile \
+		-t "${registry}/${name}:${tag}" \
```
Comment (bpradipt) [inline hack/build.sh:21]: How to distinguish the arch specific images ?

Comment (jtumber-ibm) [reply hack/build.sh:21]: buildx automatically creates a manifest list with the archs labeled e.g.

```json
{
  "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
  "schemaVersion": 2,
  "manifests": [
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "digest": "sha256:27950a58e2c2b62b569a6de1e74d5ad5b664f864df50934be873d3a9502db256",
      "size": 1091,
      "platform": {
        "architecture": "amd64",
        "os": "linux"
      }
    },
    {
      "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
      "digest": "sha256:57563ee8e32ed922819e7ead09e684d14100683c9896a7964dcbe235be73533d",
      "size": 1091,
      "platform": {
        "architecture": "s390x",
        "os": "linux"
      }
    }
  ]
}
```

Comment (bpradipt) [reply hack/build.sh:21]: nice..

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @jtumber-ibm

**Pull request merged**